README.md
# Feelings 🎭

Another rating view to share your **feelings**.

![Build Status](https://travis-ci.org/hemangshah/Feelings.svg?branch=master)
![License](https://img.shields.io/badge/License-MIT-lightgrey.svg)
![Platform](https://img.shields.io/badge/Platforms-iOS%209.0%20%E2%89%A5-red.svg)
![Swift 3.x](https://img.shields.io/badge/Swift-4.x-blue.svg)
![MadeWithLove](https://img.shields.io/badge/Made%20with%20%E2%9D%A4-India-green.svg)
[![Blog](https://img.shields.io/badge/Blog-iKiwiTech.com-blue.svg)](http://www.ikiwitech.com)
[![Awesome-Swift](https://cdn.rawgit.com/sindresorhus/awesome/d7305f38d29fed78fa85652e3a63e154dd8e8829/media/badge.svg)](https://github.com/matteocrippa/awesome-swift/)

<img src="https://github.com/hemangshah/Feelings/blob/master/Screenshots/Usage.gif"/>

## Installation

> **IMPORTANT**: You will need two images, one is filled and another is unfilled to represents feelings. See this sample images for the same, check it [here](https://github.com/hemangshah/Feelings/tree/master/Feelings/Feelings/Assets.xcassets).

1.**Manually** - Add `FeelingsView.swift` class to your Project. All set.

2.**CocoaPods**:

    source 'https://github.com/CocoaPods/Specs.git'
    target 'Sample' do
    use_frameworks!
    pod 'FeelingsView', '~>1.1'
    end
    
# Usage:

### Create Programmatically.

    //Create Sample Arrays
    let rows = ["Quality","Price","Value"]
    let columns = ["1 Star","2 Star","3 Star","4 Star","5 Star"]
        
    //Create FeelingsView
    //Note: You should provide two images for FeelingsView. 1. Filled and 2. Unfilled        
    let viewFeeling = FeelingsView.init(frame: CGRect.init(x: 0.0, y: 0.0, width: 300.0, height: 200.0))
    viewFeeling.backgroundColor = UIColor.clear
    viewFeeling.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin]
    self.view.addSubview(viewFeeling)
    viewFeeling.center = self.view.center
    
    //Setting fill/unfill images for FeelingsView
    viewFeeling.fillImage = UIImage.init(named: "filled.png")!
    viewFeeling.unfillImage = UIImage.init(named: "unfilled.png")!
        
    //Setting up values for Feelings
    viewFeeling.columnTitles = columns
    viewFeeling.rowTitles = rows 
    
    //Reload
    viewFeeling.reloadFeelingView()
        
    //Detect selection of Feelings value
    viewFeeling.onFilledCompletion = { (row,column) in
        //Note: row and column are the Int which a user tapped in the FeelingsView
        let rowValue = rows[row]
        let columnValue = columns[column]
        print("\(rowValue) -> \(columnValue)")
    }
    
> **IMPORTANT**: For customizations see the [example](https://github.com/hemangshah/Feelings/blob/master/Feelings/Feelings/ViewController.swift).

### Create in Storyboard/XIB.

1. Add a `UIView`. Set require size. Add constraints if requires.

2. Change class type from `UIView` to `FeelingsView`.
<img src="https://github.com/hemangshah/Feelings/blob/master/Screenshots/Usage-Screenshot-1.png">

3. Apply the properties for `FeelingsView`.
<img src="https://github.com/hemangshah/Feelings/blob/master/Screenshots/Usage-Screenshot-2.png">

4. Create an `IBOutlet` for `FeelingsView`. Bind it in `IBInspector`.

5. In `viewDidLoad` or at anyplace where you want provide rows and columns titles.

6. Reload `FeelingsView` by calling `reloadFeelingView` function.

7. Detect the taps on `FeelingsView` by implementing `onFilledCompletion` closure block.

# ToDos:

- [x] CocoaPods support

You can [watch](https://github.com/hemangshah/Feelings/subscription) to **Feelings** to see continuous updates. Stay tuned.

<b>Have an idea for improvements of this class?
Please open an [issue](https://github.com/hemangshah/Feelings/issues/new).</b>
    
## Credits

<b>[Hemang Shah](https://about.me/hemang.shah)</b>

**You can shoot me an [email](http://www.google.com/recaptcha/mailhide/d?k=01IzGihUsyfigse2G9z80rBw==&c=vU7vyAaau8BctOAIJFwHVbKfgtIqQ4QLJaL73yhnB3k=) to contact.**

## License

The MIT License (MIT)

> Read the [LICENSE](https://github.com/hemangshah/Feelings/blob/master/LICENSE) file for details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

Latest release is available [here](https://github.com/hemangshah/Feelings/releases/latest).

## [1.1](https://github.com/hemangshah/Feelings/releases/tag/1.1) - Jul 06, 2017
## Fixed 
- https://github.com/hemangshah/Feelings/issues/1

## [1.0](https://github.com/hemangshah/Feelings/releases/tag/1.0) - Jun 19, 2017
- Added support to install with the CocoaPods [here](https://cocoapods.org/pods/FeelingsView).


Feelings/Feelings/Feelings.swift
//
//  Feelings.swift
//  Feelings
//
//  Created by Hemang Shah on 6/14/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

let leftMarginForRows = 50.0
let topMarginForColumns = 50.0

let heightForRows = 30.0

public class FeelingsView : UIView {
    public var columnTitles = Array<String>()
    public var rowTitles = Array<String>()
    public var fillImage:UIImage?
    public var unfillImage:UIImage?
    
    init(withFrame frame:CGRect, withFillImage fImage:UIImage, withUnFillImage unFImage:UIImage) {
        super.init(frame: frame)
        self.fillImage = fImage
        self.unfillImage = unFImage
        self.backgroundColor = .white
    }
    
    fileprivate func createRows() -> Void {
        let rowLabelPointX:Double = 0.0
        var rowLabelPointY:Double = topMarginForColumns
        let feelingsViewHeight:Double = Double(self.frame.height)
        let dynamicHeight = (feelingsViewHeight - topMarginForColumns) / Double(rowTitles.count)
        
        for index in 0..<self.rowTitles.count {
            let rowLabel = createLabel(withFrame: CGRect.init(x: rowLabelPointX, y: rowLabelPointY, width: leftMarginForRows, height: dynamicHeight), text: rowTitles[index], font: UIFont.boldSystemFont(ofSize: 12.0), textColor: .black, textAlignment: .center)
            rowLabel.numberOfLines = 2
            self.addSubview(rowLabel)
            rowLabelPointY = rowLabelPointY + dynamicHeight
        }
    }
    
    fileprivate func createColumns() -> Void {
        var columnLabelPointX:Double = leftMarginForRows
        let columnLabelPointY:Double = 0.0
        let feelingsViewWidth:Double = Double(self.frame.width)
        let dynamicWidth = (feelingsViewWidth - leftMarginForRows) / Double(columnTitles.count)
        
        for index in 0..<self.columnTitles.count {
            let columnLabel = createLabel(withFrame: CGRect.init(x: columnLabelPointX, y: columnLabelPointY, width: dynamicWidth, height: topMarginForColumns), text: columnTitles[index], font: UIFont.boldSystemFont(ofSize: 12.0), textColor: .black, textAlignment: .center)
            columnLabel.numberOfLines = 2
            self.addSubview(columnLabel)
            columnLabelPointX = columnLabelPointX + dynamicWidth
        }
    }
    
    fileprivate func createFeelings() -> Void {
        var feelingsLabelPointX:Double = leftMarginForRows
        var feelingsLabelPointY:Double = topMarginForColumns
        let feelingsViewWidth:Double = Double(self.frame.width)
        let dynamicWidth = (feelingsViewWidth - leftMarginForRows) / Double(columnTitles.count)
        
        for _ in 0..<self.rowTitles.count {
            for _ in 0..<self.columnTitles.count {
                let feelingsButton = UIButton()
                feelingsButton.frame = CGRect.init(x: feelingsLabelPointX, y: feelingsLabelPointY, width: dynamicWidth, height: topMarginForColumns)
                feelingsButton.addTarget(self, action: #selector(actionFeelingsTapped), for: .touchUpInside)
                feelingsButton.setImage(unfillImage!, for: .normal)
                feelingsButton.setImage(fillImage!, for: .selected)
                feelingsButton.backgroundColor = .clear
                self.addSubview(feelingsButton)
                feelingsLabelPointX = feelingsLabelPointX + dynamicWidth
            }
            feelingsLabelPointX = leftMarginForRows
            feelingsLabelPointY = feelingsLabelPointY + topMarginForColumns
        }
    }
    
    //MARK: Actions
    @objc fileprivate func actionFeelingsTapped(feelingsButton:UIButton) -> Void {
        print("Tapped!")
        feelingsButton.isSelected = !feelingsButton.isSelected
    }
    
    public func reloadFeelingView() -> Void {
        createRows()
        createColumns()
        createFeelings()
    }
    
    //MARK: Create Label
    fileprivate func createLabel(withFrame frame:CGRect, text:String, font:UIFont, textColor:UIColor, textAlignment:NSTextAlignment) -> UILabel {
        let label = UILabel.init(frame: frame)
        label.backgroundColor = UIColor.clear
        label.text = text
        label.font = font
        label.textColor = textColor
        label.textAlignment = textAlignment
        label.minimumScaleFactor = (UIFont.labelFontSize/2)/UIFont.labelFontSize
        label.adjustsFontSizeToFitWidth = true
        return label
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}


Feelings/Feelings/AppDelegate.swift
//
//  AppDelegate.swift
//  Feelings
//
//  Created by Hemang Shah on 6/14/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the active state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Feelings/Feelings/FeelingsView.swift
//
//  FeelingsView.swift
//  Feelings
//
//  Created by Hemang Shah on 6/14/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

//Credits for the Animation: [SeanAllen] https://gist.github.com/SAllen0400/a09754049fcdcc00695291b3a011fbbd

fileprivate extension UIButton {
    
    func pulsate() {
        
        let pulse = CASpringAnimation(keyPath: "transform.scale")
        pulse.duration = 0.2
        pulse.fromValue = 0.95
        pulse.toValue = 1.0
        pulse.autoreverses = true
        pulse.repeatCount = 2
        pulse.initialVelocity = 0.5
        pulse.damping = 1.0
        
        layer.add(pulse, forKey: "pulse")
    }
    
    func flash() {
        
        let flash = CABasicAnimation(keyPath: "opacity")
        flash.duration = 0.2
        flash.fromValue = 1
        flash.toValue = 0.1
        flash.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        flash.autoreverses = true
        flash.repeatCount = 3
        
        layer.add(flash, forKey: nil)
    }
    
    
    func shake() {
        
        let shake = CABasicAnimation(keyPath: "position")
        shake.duration = 0.05
        shake.repeatCount = 2
        shake.autoreverses = true
        
        let fromPoint = CGPoint(x: center.x - 5, y: center.y)
        let fromValue = NSValue(cgPoint: fromPoint)
        
        let toPoint = CGPoint(x: center.x + 5, y: center.y)
        let toValue = NSValue(cgPoint: toPoint)
        
        shake.fromValue = fromValue
        shake.toValue = toValue
        
        layer.add(shake, forKey: "position")
    }
}

//You can change this value accordingly and it will update the entire FeelingsView
fileprivate let leftMarginForRows = 50.0
fileprivate let topMarginForColumns = 50.0

public enum FeelingsViewButtonTapAnimationType: Int {
    case None
    case Pulse
    case Flash
    case Shake
}

@IBDesignable
public class FeelingsView : UIView {
    
    ///columnTitles of Type Array<String>
    public var columnTitles = Array<String>()
    ///rowTitles of Type Array<String>
    public var rowTitles = Array<String>()
    ///Animation Type for the Feelings Button Taps.
    public var feelingsButtonAnimationType: FeelingsViewButtonTapAnimationType = .None
    ///Default Feelings Value.
    public var defaultFeeligns:Int = -1
    
    //Fonts
    @IBInspectable public var rowTitleFont:UIFont?
    @IBInspectable public var columnTitleFont:UIFont?
    
    //Colors
    @IBInspectable public var rowTitleColor:UIColor?
    @IBInspectable public var columnTitleColor:UIColor?
    
    //Background Colors
    @IBInspectable public var rowTitleBackgroundColor:UIColor?
    @IBInspectable public var columnTitleBackgroundColor:UIColor?
    @IBInspectable public var feelingsButtonsBackgroundColor:UIColor?
    
    //fill Image (Uses to show radio button selection)
    @IBInspectable public var fillImage:UIImage? = nil
    //unfillImage Image (Uses to show radio button not selected)
    @IBInspectable public var unfillImage:UIImage? = nil
    
    //Completion Block to Detect selection of Feelings value (row,column).
    public var onFilledCompletion:((_ row:Int, _ column:Int) -> ())? = nil
    
    //MARK:Init
    public override init(frame: CGRect) {
        super.init(frame: frame)
        //By default, you can customize it to your wish.
        self.backgroundColor = .clear
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    //MARK: Create Rows/Columns/Feelings
    fileprivate func createRows() -> Void {
        let rowLabelPointX:Double = 0.0
        var rowLabelPointY:Double = topMarginForColumns
        let feelingsViewHeight:Double = Double(self.frame.height)
        let dynamicHeight = (feelingsViewHeight - topMarginForColumns) / Double(rowTitles.count)
        
        for index in 0..<self.rowTitles.count {
            let rowLabel = createLabel(withFrame: CGRect.init(x: rowLabelPointX, y: rowLabelPointY, width: leftMarginForRows, height: dynamicHeight), text: rowTitles[index], font: UIFont.boldSystemFont(ofSize: 12.0), textColor: .black, textAlignment: .center)
            rowLabel.numberOfLines = 2
            if let rowTitleColor = rowTitleColor {
                rowLabel.textColor = rowTitleColor
            }
            if let rowTitleBackgroundColor = rowTitleBackgroundColor {
                rowLabel.backgroundColor = rowTitleBackgroundColor
            }
            if let rowTitleFont = rowTitleFont {
                rowLabel.font = rowTitleFont
            }
            self.addSubview(rowLabel)
            rowLabelPointY = rowLabelPointY + dynamicHeight
        }
    }
    
    fileprivate func createColumns() -> Void {
        var columnLabelPointX:Double = leftMarginForRows
        let columnLabelPointY:Double = 0.0
        let feelingsViewWidth:Double = Double(self.frame.width)
        let dynamicWidth = (feelingsViewWidth - leftMarginForRows) / Double(columnTitles.count)
        
        for index in 0..<self.columnTitles.count {
            let columnLabel = createLabel(withFrame: CGRect.init(x: columnLabelPointX, y: columnLabelPointY, width: dynamicWidth, height: topMarginForColumns), text: columnTitles[index], font: UIFont.boldSystemFont(ofSize: 12.0), textColor: .black, textAlignment: .center)
            columnLabel.numberOfLines = 2
            if let columnTitleColor = columnTitleColor {
                columnLabel.textColor = columnTitleColor
            }
            if let columnTitleBackgroundColor = columnTitleBackgroundColor {
                columnLabel.backgroundColor = columnTitleBackgroundColor
            }
            if let columnTitleFont = columnTitleFont {
                columnLabel.font = columnTitleFont
            }
            self.addSubview(columnLabel)
            columnLabelPointX = columnLabelPointX + dynamicWidth
        }
    }
    
    fileprivate func createFeelings() -> Void {
        var feelingsViewPointX:Double = leftMarginForRows
        var feelingsViewPointY:Double = topMarginForColumns
        let feelingsViewWidth:Double = Double(self.frame.width)
        let dynamicWidth = (feelingsViewWidth - leftMarginForRows) / Double(columnTitles.count)
        
        for rowIndex in 0..<self.rowTitles.count {
            
            let rowView = UIView.init(frame: CGRect.init(x: feelingsViewPointX, y: feelingsViewPointY, width: dynamicWidth * Double(columnTitles.count), height: topMarginForColumns))
            rowView.tag = (rowIndex + 1)
            rowView.backgroundColor = .clear
            self.addSubview(rowView)
            
            var feelingsButtonPointX:Double = 0.0
            for columnIndex in 0..<self.columnTitles.count {
                let feelingsButton = UIButton()
                feelingsButton.tag = columnIndex + 1
                feelingsButton.frame = CGRect.init(x: feelingsButtonPointX, y: 0.0, width: dynamicWidth, height: topMarginForColumns)
                feelingsButton.addTarget(self, action: #selector(actionFeelingsTapped), for: .touchUpInside)
                feelingsButton.setImage(unfillImage!, for: .normal)
                feelingsButton.setImage(fillImage!, for: .selected)
                feelingsButton.backgroundColor = .clear
                if let feelingsButtonsBackgroundColor = feelingsButtonsBackgroundColor {
                    feelingsButton.backgroundColor = feelingsButtonsBackgroundColor
                }
                rowView.addSubview(feelingsButton)
                feelingsButtonPointX = feelingsButtonPointX + dynamicWidth
                if defaultFeeligns >= 1 && defaultFeeligns <= columnTitles.count {
                    if defaultFeeligns == (columnIndex + 1) {
                        actionFeelingsTapped(button: feelingsButton)
                    }
                }
            }
            feelingsViewPointX = leftMarginForRows
            feelingsViewPointY = feelingsViewPointY + topMarginForColumns
        }
    }
    
    //MARK: Actions
    @objc fileprivate func actionFeelingsTapped(button:UIButton) -> Void {
        if feelingsButtonAnimationType != .None {
            if feelingsButtonAnimationType == .Pulse {
                button.pulsate()
            } else if feelingsButtonAnimationType == .Flash {
                button.flash()
            } else if feelingsButtonAnimationType == .Shake {
                button.shake()
            }
        }
        
        let rowView = button.superview!
        unFilledEveryOneInRow(button: button, withRowView: rowView)
        button.isSelected = !button.isSelected
        button.isUserInteractionEnabled = !button.isUserInteractionEnabled
        if onFilledCompletion != nil {
            onFilledCompletion!(rowView.tag-1, button.tag-1)
        }
    }
    
    //MARK: Helpers
    fileprivate func unFilledEveryOneInRow(button:UIButton, withRowView rowView:UIView) -> Void {
        for case let feelingsButton as UIButton in rowView.subviews {
            feelingsButton.isSelected = false
            feelingsButton.isUserInteractionEnabled = true
        }
    }
    
    //MARK: Create Label
    fileprivate func createLabel(withFrame frame:CGRect, text:String, font:UIFont, textColor:UIColor, textAlignment:NSTextAlignment) -> UILabel {
        let label = UILabel.init(frame: frame)
        label.backgroundColor = UIColor.clear
        label.text = text
        label.font = font
        label.textColor = textColor
        label.textAlignment = textAlignment
        label.minimumScaleFactor = (UIFont.labelFontSize/2)/UIFont.labelFontSize
        label.adjustsFontSizeToFitWidth = true
        return label
    }
    
    //MARK: Reload Feelings View
    ///FeelingsView should be reload after setting rows and columns titles.
    public func reloadFeelingView() -> Void {
        guard fillImage != nil || unfillImage != nil else {
            print("You forgot to provide fillImage or unfillImage.")
            return
        }
        
        guard rowTitles.count > 0 && columnTitles.count > 0 else {
            print("You forgot to provide rowTitles or columnTitles values.")
            return
        }
        createRows()
        createColumns()
        createFeelings()
    }
}


Feelings/Feelings/ViewController.swift
//
//  ViewController.swift
//  Feelings
//
//  Created by Hemang Shah on 6/14/17.
//  Copyright © 2017 Hemang Shah. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        //Create Sample Arrays
        let rows = ["Quality","Price","Value"]
        let columns = ["1 Star","2 Star","3 Star","4 Star","5 Star"]
        
        //Create FeelingsView
        //Note: You should provide two images for FeelingsView. 1. Filled and 2. Unfilled
        let viewFeeling = FeelingsView.init(frame: CGRect.init(x: 0.0, y: 0.0, width: 300.0, height: 200.0))
        viewFeeling.backgroundColor = UIColor.clear
        viewFeeling.autoresizingMask = [.flexibleLeftMargin, .flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin]
        self.view.addSubview(viewFeeling)
        viewFeeling.center = self.view.center
        
        //Setting fill/unfill images for FeelingsView
        viewFeeling.fillImage = UIImage.init(named: "filled.png")!
        viewFeeling.unfillImage = UIImage.init(named: "unfilled.png")!
        
        //Setting up values for FeelingsView
        viewFeeling.columnTitles = columns
        viewFeeling.rowTitles = rows
        
        //Add animations when feelings button tapped.
        //Pulse, Flash, Shake and None.
        viewFeeling.feelingsButtonAnimationType = .Flash
        
        //Default Feelings Value (from 1 up-to total columns count)
        viewFeeling.defaultFeeligns = 5
        
        //Customization Options
        //viewFeeling.backgroundColor = .red
        
        //viewFeeling.rowTitleColor = .white
        //viewFeeling.columnTitleColor = .white
        
        //viewFeeling.rowTitleFont = UIFont.init(name: "Verdana", size: 14.0)
        //viewFeeling.columnTitleFont = UIFont.init(name: "Helvetica", size: 14.0)
        
        //viewFeeling.rowTitleBackgroundColor = .red
        //viewFeeling.columnTitleBackgroundColor = .red
        
        //viewFeeling.feelingsButtonsBackgroundColor = .white
        
        //Reload
        viewFeeling.reloadFeelingView()
        
        //Detect selection of Feelings value
        viewFeeling.onFilledCompletion = { (row,column) in
            //Note: row and column are the Int which a user tapped in the FeelingsView
            let rowValue = rows[row]
            let columnValue = columns[column]
            print("\(rowValue) -> \(columnValue)")
        }
    }
}
